src/ctx.cpp

```diff
@@ -3436,20 +3436,26 @@ void FunctionEmitContext::MemcpyInst(llvm::Value *dest, llvm::Value *src, llvm::
 #endif
 }
 
-void FunctionEmitContext::setLoopUnrollMetadata(llvm::Instruction *inst,
-                                                std::pair<Globals::pragmaUnrollType, int> loopAttribute,
-                                                SourcePos pos) {
+void FunctionEmitContext::setLoopMetadata(llvm::Instruction *inst,
+                                          std::pair<Globals::pragmaUnrollType, int> loopAttribute, SourcePos pos,
+                                          bool hasConstantCondition) {
     if (inst == nullptr) {
         return;
     }
 
-    if (loopAttribute.first == Globals::pragmaUnrollType::none) {
+    // Always create metadata if we need to add mustprogress, even if there's no unroll pragma
+    bool hasUnrollMetadata = (loopAttribute.first != Globals::pragmaUnrollType::none);
```

Comment:
Style: It's probably better to try avoiding double negation and go with straightforward logic for UnrollMetadata: `noUnrollMetadata`. That would be easier to understand the logic.

For `shouldAddMustProgress` I would name it `needMustProgress`

Feel free to dismiss this if you prefer.